Work-in-progress

tgross opened the pull request:
@misterbisson looking for your comments here.

tgross committed:
Project skeleton and initial implementation

Included at this point:
- Dockerfile for building Touchbase and including Containerbuddy
- docker-compose files for Triton and local testing
- Dockerfile for Nginx (mostly lifted from Containerbuddy example app)
- startup script that starts all containers and configures CB tables

misterbisson commented on README.md:
```diff
@@ -1,2 +1,4 @@
-# triton-touchbase
-Touchbase stack designed for container-native deployment on Joyent's Triton platform
+triton-elasticsearch
```
Elastic?

Separately, I don't know if it's significant that the license in this repo is different from [the license in the Touchbase repo](https://github.com/couchbaselabs/touchbase/blob/master/LICENSE.md).

tgross replied:
Oops, copypasta problem.

We're not distributing the Touchbase code as our own in this case, merely installing it as part of our application. At most we might include the license for Touchbase within the LICENSE file.

tgross commented:
@misterbisson I've rebased some of the WIP commits, cleaned up some vague language in comments and corrected that phrase in the README. At this point I mostly just need to properly wire up `onChange` handlers.

tgross committed:
Write templates to Consul during Nginx and Touchbase first-boot

tgross committed:
Update Nginx virtualhost config and template to have only TB upstream

tgross committed:
Update file path for TB template

tgross committed:
Fixed template path, make createBucket wait

tgross committed:
Update Touchbase templates, add run-touchbase.sh script

tgross committed:
Add .env file creation to start.sh script

tgross commented:
An issue that's going to be a bit of a pain here is the SendGrid integration. Signing up for a SendGrid account is a total pain in the ass because they want to do things like domain-verify you. If we put that barrier in the way of people trying this out they're going to be annoyed.

My proposal to work-around this is to just patch-out the requirement for SendGrid API access by updating this line https://github.com/couchbaselabs/touchbase/blob/adf73c2980ef2f33e6c7a61e14a583f9e6aa0040/TouchbaseModular/models/usermodel.js#L90 to read `emailVerified: true`. This way new users are automatically treated as verified and can log in.

We'll make sure to note that we're doing this in the blog post and README, of course.

misterbisson commented:
> patch-out the requirement for SendGrid API access by updating this line

Sounds like a fine plan to me. Is that an argument for forking the source repo, though?

tgross commented:
> Sounds like a fine plan to me. Is that an argument for forking the source repo, though?

Might be easier to do. This way we can pin the code to a specific version and not worry about the developer coming along and breaking the blueprint.

tgross commented:
Did that patch-out in https://github.com/tgross/touchbase/tree/no_sendgrid and that's what we'll clone in the Dockerfile.

tgross committed:
Remove sendgrid requirement for signup

tgross committed:
Fixed a JSON error in the TB template

tgross committed:
Make sure images have shipped to Triton and that docker-compose expects the correct ones

tgross committed:
N1QL port needs to be exposed for Couchbase

tgross commented:
Code-wise this should be in good shape and has been tested end-to-end. I need to work up the README and I'm going to do that in parallel with writing up the blog post, which I've opened a PR for at https://github.com/joyent/blog/pull/61

tgross committed:
Update start.sh script to match blog post

tgross committed:
Update README to match a (trimmed-down) version of the blog post

tgross committed:
Change Nginx load balancing to least_conns

tgross committed:
Make scale-up example match diagram

misterbisson commented on README.md:
```diff
@@ -1,2 +1,94 @@
+
+![Diagram of Touchbase architecture](./doc/triton-touchbase.png)
+
+### Couchbase
```
I'm a bit tempted to remove the list of components as a way to minimize the duplication of content between this and the blog post.

tgross replied:
Yeah, that's something I hadn't considered.

misterbisson replied:
Strongly related: https://github.com/tgross/triton-touchbase/pull/1#discussion_r44156783

misterbisson commented on README.md:
```diff
@@ -1,2 +1,94 @@
+
+This repo is a demonstation of a multi-tier container-native application architecture on [Joyent's](https://www.joyent.com/) Triton platform. It uses Touchbase and its supporting services as an example. It includes the following components:
+
+- Consul, acting as a discovery service
```
How would you feel about linking these to the blog post, anchored on the relevant section? Ex:

> - [Consul](https://example.com/blog/putting-it-all-together#couchbase), acting as a discovery service

misterbisson replied:
Obviously, this would have to be done in a followup commit, after we know the proper URLs.

tgross replied:
And then just remove the subsections between here and "running the example"? Sounds good to me.

tgross committed:
Updated Cloudflare to CloudFlare in diagram

tgross committed:
Remove duplicate content from README that would conflict with blog post

misterbisson commented on start.sh:
```diff
@@ -0,0 +1,243 @@
+startDatabase() {
+    echo
+    echo 'Starting Couchbase'
+    ${COMPOSE} up -d --no-recreate couchbase
```
Is there a reason to start each component separately, rather than letting Compose start everything at once and letting them reconfigure as their upstreams appear?

misterbisson replied:
I should at that I'm thinking about this now as I'm working through the demo talk track for it and thinking about how provisioning we can parallelize provisioning, startup, and configuration times.

tgross replied:
If we don't start Couchbase first we're at risk of a race condition between the first node of Couchbase being started and registered with Consul, and the first Touchbase node coming up. The first CB node needs to be up and registered because we're relying on the CB client to be able to access it (because TB doesn't support graceful reload).

misterbisson replied:
> The first CB node needs to be up and registered because we're relying on the CB client to be able to access it (because TB doesn't support graceful reload).

That's a huge use-case for our conversation about an executable that runs before the primary application runs, yes?

tgross replied:
> That's a huge use-case for our conversation about an executable that runs before the primary application runs, yes?

Absolutely. Will be fleshed out in https://github.com/joyent/containerbuddy/issues/13

misterbisson commented on .env.example:
```diff
@@ -0,0 +1,10 @@
+SERVICE=
+RECORD=
+TTL=
+CB_USER=
```
Two points worth considering:
1. The Couchbase container expects `COUCHBASE_USER`. 
2. `CB` seems ambiguous, especially since we launched Containerbuddy.

Related: https://github.com/tgross/triton-touchbase/issues/2

misterbisson commented on docker-compose.yml:
```diff
@@ -0,0 +1,82 @@
+        - 18092
+    environment:
+        - CONSUL_HOST=http://consul:8500
+        - COUCHBASE_SERVICE_NAME=couchbase
```
See https://github.com/tgross/triton-touchbase/issues/2

misterbisson commented on .env.example:
```diff
@@ -0,0 +1,10 @@
+TTL=
+CB_USER=
+CB_PASSWORD=
+SENDGRID_USER=
```
See https://github.com/tgross/triton-touchbase/issues/4

misterbisson commented on start.sh:
```diff
@@ -0,0 +1,243 @@
+
+# send a REST API call to remove a CB bucket
+removeBucket() {
+    curl -X DELETE -vvv http://${CBAPI}/pools/default/buckets/$1 \
```
See https://github.com/tgross/triton-touchbase/issues/3

misterbisson commented on start.sh:
```diff
@@ -0,0 +1,243 @@
+# send a REST API call to Couchbase to create a N1QL index
+createIndex() {
+    echo $1
+    curl -s --fail -s -X POST http://${N1QLAPI}/query/service \
```
See https://github.com/tgross/triton-touchbase/issues/3

misterbisson commented on README.md:
````diff
@@ -1,2 +1,74 @@
+
+```
+Touchbase cluster is launched!
+Try scaling it up by running: ./start scale
````
`./start.sh`, not `./start`

misterbisson commented on start.sh:
```diff
@@ -0,0 +1,243 @@
+
+echo
+echo 'Touchbase cluster is launched!'
+echo "Try scaling it up by running: ./start ${CONFIG_FILE} scale"
```
`./start.sh`, not `./start`. Also, do you intend to include `${CONFIG_FILE}`?

misterbisson commented on start.sh:
```diff
@@ -0,0 +1,243 @@
+    N1QLAPI=$(getIpPort couchbase 8093)
+    echo
+    echo 'Creating Couchbase buckets'
+    removeBucket benchmark
```
I get a failure on this semi-regularly, though it might be the bad wifi:

```
Creating Couchbase buckets
*   Trying 165.225.159.37...
* Connected to 165.225.159.37 (165.225.159.37) port 8091 (#0)
* Server auth using Basic with user 'Administrator'
> DELETE /pools/default/buckets/benchmark HTTP/1.1
> Host: 165.225.159.37:8091
> Authorization: Basic QWRtaW5pc3RyYXRvcjpwYXNzd29yZA==
> User-Agent: curl/7.43.0
> Accept: */*
> 
< HTTP/1.1 404 Object Not Found
< Server: Couchbase Server
< Pragma: no-cache
< Date: Mon, 16 Nov 2015 10:01:50 GMT
< Content-Type: text/plain
< Content-Length: 31
< Cache-Control: no-cache
< 
Requested resource not found.
* Connection #0 to host 165.225.159.37 left intact
..SUCCESS: bucket-create
...SUCCESS: bucket-create
HTTP/1.1 500 Connection Not Upgraded
Connection: close
Date: Mon, 16 Nov 2015 10:02:35 GMT

Creating Couchbase indexes
CREATE PRIMARY INDEX ON users
```

misterbisson commented on README.md:
```diff
@@ -1,2 +1,74 @@
+CF_ROOT_DOMAIN=<the root domain you want to manage. ex. example.com>
+SERVICE=nginx <the name of the service you want to monitor>
+RECORD=<the A-record you want to manage. ex. my.example.com>
+TTL=600 <the DNS TTL you want>
```
> Must be between 120 and 2,147,483,647 seconds

tgross replied:
It's between 1 and that, per the CloudFlare docs. Fixed in https://github.com/tgross/triton-touchbase/commit/4d9af596f50b4e8327ab3e827c3218060ba090db
